embassy-stm32/src/timer/qei.rs

```diff
@@ -32,16 +32,29 @@ pub struct QeiPin<'d, T, Channel> {
 macro_rules! channel_impl {
     ($new_chx:ident, $channel:ident, $pin_trait:ident) => {
         impl<'d, T: GeneralInstance4Channel> QeiPin<'d, T, $channel> {
-            #[doc = concat!("Create a new ", stringify!($channel), " QEI pin instance.")]
-            pub fn $new_chx(pin: impl Peripheral<P = impl $pin_trait<T>> + 'd) -> Self {
-                into_ref!(pin);
-                critical_section::with(|_| {
-                    pin.set_low();
-                    pin.set_as_af(pin.af_num(), AfType::input(Pull::None));
-                });
-                QeiPin {
-                    _pin: pin.map_into(),
-                    phantom: PhantomData,
+            paste! {
+                fn [<_ $new_chx>](pin: impl Peripheral<P = impl $pin_trait<T>> + 'd, pull: Pull) -> Self {
+                    into_ref!(pin);
+                    critical_section::with(|_| {
+                        pin.set_low();
+                        pin.set_as_af(pin.af_num(), AfType::input(pull));
+                    });
+                    QeiPin {
+                        _pin: pin.map_into(),
+                        phantom: PhantomData,
+                    }
+                }
+                #[doc = concat!("Create a new ", stringify!($channel), " QEI pin instance.")]
+                pub fn $new_chx(pin: impl Peripheral<P = impl $pin_trait<T>> + 'd) -> Self {
+                    Self::[<_ $new_chx>](pin, Pull::None)
+                }
+                #[doc = concat!("Create a new ", stringify!($channel), " QEI pin instance with internal pull-up.")]
+                pub fn [<$new_chx _pu>](pin: impl Peripheral<P = impl $pin_trait<T>> + 'd) -> Self {
+                    Self::[<_ $new_chx>](pin, Pull::Up)
+                }
+                #[doc = concat!("Create a new ", stringify!($channel), " QEI pin instance with internal pull-down.")]
+                pub fn [<$new_chx _pd>](pin: impl Peripheral<P = impl $pin_trait<T>> + 'd) -> Self {
```

Comment:
instead of making three functions, could you add a `Config` struct with the pull as a field? This way it's consistent with how other drivers do it (for example [usart](https://github.com/embassy-rs/embassy/blob/main/embassy-stm32/src/usart/mod.rs#L208)) and will hopefully let us avoid `paste` for a little longer.

Reply:
That'll break the API I think unless I can figure out how to do optional args. But I'll sure take a look. Thanks for the fast feedback.

Reply:
no need, doing breaking changes is fine if needed, and in this case it's what will result in a simpler API and more consistent with the rest of embassy-stm32.

(a nice thing about `#[non_exhaustive] struct Config` is that after we've added it we can add *other* config fields without it being breaking.)